### PR TITLE
test: fix zmq test flakiness, improve speed

### LINKED ISSUE #20934: Fix zmq test flakiness
There are many reports of the test being flaky: https://github.com/bitcoin/bitcoin/issues/20672#issuecomment-746260431

Thus, it should be made more robust, as described in https://github.com/bitcoin/bitcoin/issues/20538#issuecomment-738791868

#### Useful skills:

* Background in our functional test suite (python3)
* Background in zmq

#### Want to work on this issue?

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by theStack**
Fixes #20934 by using the "sync up" method described in https://github.com/bitcoin/bitcoin/issues/20538#issuecomment-738791868.

After improving robustness with this approach (commits 1-3), it turned out that there were still some fails, but those were unrelated to zmq: Out of 500 runs, 3 times `sync_mempool()` or `sync_blocks()` timed out, which can happen because the trickle relay time has no upper bound -- hence in rare cases, it takes longer than 60s. This is fixed by enabling immediate tx relay on node1 (commit 4), which as a nice side-effect also gives us a rough 2x speedup for the test.

For further details, also see the explanations in the commit messages.

There is no guarantee that the test is still not flaky, but it would help if potential reviewers would run the following script locally and report how many runs failed (feel free to do less than 1000 runs, as this takes quite a long if ran with `--valgrind`):
```
#!/bin/sh
OUTPUT_FILE=./zmq_results
echo ===== repeated zmq test ===== > $OUTPUT_FILE

for i in `seq 1000`; do
    echo ------------------------
    echo ----- test run $i -----
    echo ------------------------
    echo --- $i --- >> $OUTPUT_FILE
    ./test/functional/interface_zmq.py --valgrind
    if [ $? -ne 0 ]; then
        echo "FAILED. /o\\" >> $OUTPUT_FILE
    else
        echo "PASSED. \\o/" >> $OUTPUT_FILE
    fi
done

echo Failed test runs:
grep FAILED $OUTPUT_FILE | wc -l
```

### CODE REVIEW

**Comment by theStack** [comment]
Asked to mark this as draft since the test now fails on two cirrus instances. Have to investigate deeper what the problem is and how to properly reproduce it... locally 1000 test runs passed successfully.

**Comment by MarcoFalke** [comment]
You might want to try with `--valgrind`, which makes races more likely to happen locally

**Comment by theStack** [comment]
@MarcoFalke: Thanks, that helped a lot! On master, running via `--valgrind` leads to a failed test run quite quickly on my machine.

The PR is ready for review now. The problem of my original approach was that most tests assumed that the chain need nodes 0 and 1 to have the same tip. So, after this robust "sync up" setup of approach with generating blocks, node 1 has to catch up. Added a parameter "sync_blocks" for that purpose, as for the last test, the chains are already different and synchronization is not possible in a trivial way. Now, I had locally a few hundred test runs with `--valgrind` and all of them passed.
Also adapted the PR description to include a script, that reviews can use to test the robustness.

**Comment by jonatack** [comment]
Concept ACK, will be great to robustify this test.

So far your script has run the test 25 times without errors for me (other than an unrelated issue that the test runner `--valgrind` option raises for me in general (not just on this test), but `valgrind test/functional/interface_zmq.py` works fine).

<details><summary>unrelated valgrind error output</summary><p>

```
------------------------
----- test run 1 -----
------------------------
2021-01-28T17:22:07.128000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_awm_44bl
2021-01-28T17:22:34.997000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 108, in _request
    return self._get_response()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 168, in _get_response
    http_response = self.__conn.getresponse()
  File "/usr/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 125, in main
    self.setup()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 262, in setup
    self.setup_network()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 356, in setup_network
    self.setup_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 383, in setup_nodes
    self.import_deterministic_coinbase_privkeys()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 400, in import_deterministic_coinbase_privkeys
    self.init_wallet(i)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 407, in init_wallet
    n.createwallet(wallet_name=wallet_name, descriptors=self.options.descriptors, load_on_startup=True)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 676, in createwallet
    return self.__getattr__('createwallet')(wallet_name, disable_private_keys, blank, passphrase, avoid_reuse, descriptors, load_on_startup)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 144, in __call__
    response, status = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 113, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python3.9/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/usr/lib/python3.9/http/client.py", line 950, in send
    self.connect()
  File "/usr/lib/python3.9/http/client.py", line 921, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.9/socket.py", line 843, in create_connection
    raise err
  File "/usr/lib/python3.9/socket.py", line 831, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused
2021-01-28T17:22:35.049000Z TestFramework (INFO): Stopping nodes
2021-01-28T17:22:35.050000Z TestFramework.node0 (ERROR): Unable to stop node.
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 320, in stop_node
    self.stop(wait=wait)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 144, in __call__
    response, status = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 107, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python3.9/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1266, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python3.9/http/client.py", line 1092, in putrequest
    raise CannotSendRequest(self.__state)
http.client.CannotSendRequest: Request-sent
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/./test/functional/interface_zmq.py", line 529, in <module>
    ZMQTest().main()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 149, in main
    exit_code = self.shutdown()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 278, in shutdown
    self.stop_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 525, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 334, in stop_node
    raise AssertionError("Unexpected stderr {} != {}".format(stderr, expected_stderr))
AssertionError: Unexpected stderr ==37264== Thread 10 b-httpworker.3:
==37264== Conditional jump or move depends on uninitialised value(s)
==37264==    at 0xB0DEC5: __log_putr.isra.2 (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xB0F090: __log_put (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xB99C1D: __crdel_metasub_log (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xAD526A: __db_log_page (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xB4068A: __bam_new_subdb (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xAE7A6D: __db_init_subdb (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xB044ED: __fop_subdb_setup (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xAE6EB0: __db_open (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xAE1719: __db_open_pp (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0xAC1D3A: Db::open(DbTxn*, char const*, char const*, DBTYPE, unsigned int, int) (in /home/jon/projects/bitcoin/bitcoin/src/bitcoind)
==37264==    by 0x873618: BerkeleyDatabase::Open() (bdb.cpp:345)
==37264==    by 0x873184: BerkeleyBatch::BerkeleyBatch(BerkeleyDatabase&, bool, bool) (bdb.cpp:308)
==37264== 
{
   <insert_a_suppression_name_here>
   Memcheck:Cond
   fun:__log_putr.isra.2
   fun:__log_put
   fun:__crdel_metasub_log
   fun:__db_log_page
   fun:__bam_new_subdb
   fun:__db_init_subdb
   fun:__fop_subdb_setup
   fun:__db_open
   fun:__db_open_pp
   fun:_ZN2Db4openEP5DbTxnPKcS3_6DBTYPEji
   fun:_ZN16BerkeleyDatabase4OpenEv
   fun:_ZN13BerkeleyBatchC1ER16BerkeleyDatabasebb
}
==37264== 
==37264== Exit program on first error (--exit-on-first-error=yes) != 
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process
```
</p></details>

**Review by jonatack — COMMENTED** [review]
Almost-ACK. I was unable to make the test fail with valgrind both here and on master. The test does run ~2x faster on this branch, which is great.

File: test/functional/interface_zmq.py [hunk 1]
```diff
@@ -82,23 +87,46 @@ def run_test(self):
...
+                break
+
+        # set subscriber's desired timeout for the test
+        for sub in subscribers:
+            sub.socket.set(zmq.RCVTIMEO, recv_timeout*1000)
+
+        if sync_blocks:
+            self.connect_nodes(0, 1)
+            self.sync_blocks()
+            self.disconnect_nodes(0, 1)
```
[the remark this answers is not in the document]
Comment (theStack) [reply test/functional/interface_zmq.py]: Agree that this was unnecessarily complicated. Changed the setup routine to _always_ connect nodes 0 and 1 (which is needed for the block sync anways, that is used in every subtest except the last one) and disconnect in the subtest if necessary.

File: test/functional/interface_zmq.py [hunk 2]
```diff
@@ -62,6 +65,8 @@ def receive_sequence(self):
 class ZMQTest (BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
+        # immediate tx relay (node1 -> node0)
+        self.extra_args = [[], ["-whitelist=noban@127.0.0.1"]]
```
[the remark this answers is not in the document]
Comment (theStack) [reply test/functional/interface_zmq.py]: Thanks, I like the clear explanation and used it. My initial reason for not using the whitelist parameter on all nodes was that for node0 they would be overwritten in the setup routine. But simply adding `self.extra_args[0]` to the extra_args parameter for the `restart_node` call also tackles this.

**Comment by theStack** [comment]
Force-pushed with changes suggested by @jonatack (https://github.com/bitcoin/bitcoin/pull/21008#discussion_r566331714 and https://github.com/bitcoin/bitcoin/pull/21008#discussion_r566336615).

File: test/functional/interface_zmq.py [hunk 3]
```diff
@@ -82,23 +88,46 @@ def run_test(self):
 
     # Restart node with the specified zmq notifications enabled, subscribe to
     # all of them and return the corresponding ZMQSubscriber objects.
-    def setup_zmq_test(self, services, recv_timeout=60, connect_nodes=False):
+    def setup_zmq_test(self, services, recv_timeout=60, sync_blocks=True):
```
Comment (jonatack) [inline test/functional/interface_zmq.py:91]: nit, can enforce named args with 
```suggestion
    def setup_zmq_test(self, services, *, recv_timeout=60, sync_blocks=True):
```
or alternatively for all args
```diff
@@ -88,7 +88,7 @@ class ZMQTest (BitcoinTestFramework):
 
     # Restart node with the specified zmq notifications enabled, subscribe to
     # all of them and return the corresponding ZMQSubscriber objects.
-    def setup_zmq_test(self, services, recv_timeout=60, sync_blocks=True):
+    def setup_zmq_test(self, *, services, recv_timeout=60, sync_blocks=True):
         subscribers = []
         for topic, address in services:
             socket = self.ctx.socket(zmq.SUB)
@@ -137,7 +137,7 @@ class ZMQTest (BitcoinTestFramework):
         self.restart_node(0, ["-zmqpubrawtx=foo", "-zmqpubhashtx=bar"])
 
         address = 'tcp://127.0.0.1:28332'
-        subs = self.setup_zmq_test([(topic, address) for topic in ["hashblock", "hashtx", "rawblock", "rawtx"]])
+        subs = self.setup_zmq_test(services=[(topic, address) for topic in ["hashblock", "hashtx", "rawblock", "rawtx"]])
 
         hashblock = subs[0]
         hashtx = subs[1]
@@ -212,7 +212,7 @@ class ZMQTest (BitcoinTestFramework):
 
         # Should only notify the tip if a reorg occurs
         hashblock, hashtx = self.setup_zmq_test(
-            [(topic, address) for topic in ["hashblock", "hashtx"]],
+            services=[(topic, address) for topic in ["hashblock", "hashtx"]],
             recv_timeout=2)  # 2 second timeout to check end of notifications
         self.disconnect_nodes(0, 1)
 
@@ -262,7 +262,7 @@ class ZMQTest (BitcoinTestFramework):
         <32-byte hash>A<8-byte LE uint> : Transactionhash added mempool
         """
         self.log.info("Testing 'sequence' publisher")
-        [seq] = self.setup_zmq_test([("sequence", "tcp://127.0.0.1:28333")])
+        [seq] = self.setup_zmq_test(services=[("sequence", "tcp://127.0.0.1:28333")])
         self.disconnect_nodes(0, 1)
 
         # Mempool sequence number starts at 1
@@ -414,7 +414,7 @@ class ZMQTest (BitcoinTestFramework):
             return
 
         self.log.info("Testing 'mempool sync' usage of sequence notifier")
-        [seq] = self.setup_zmq_test([("sequence", "tcp://127.0.0.1:28333")])
+        [seq] = self.setup_zmq_test(services=[("sequence", "tcp://127.0.0.1:28333")])
 
         # In-memory counter, should always start at 1
         next_mempool_seq = self.nodes[0].getrawmempool(mempool_sequence=True)["mempool_sequence"]
@@ -514,7 +514,7 @@ class ZMQTest (BitcoinTestFramework):
 
     def test_multiple_interfaces(self):
         # Set up two subscribers with different addresses
-        subscribers = self.setup_zmq_test([
+        subscribers = self.setup_zmq_test(services=[
             ("hashblock", "tcp://127.0.0.1:28334"),
             ("hashblock", "tcp://127.0.0.1:28335"),
         ], sync_blocks=False)
```

File: test/functional/interface_zmq.py [hunk 4]
```diff
@@ -488,7 +517,7 @@ def test_multiple_interfaces(self):
         subscribers = self.setup_zmq_test([
             ("hashblock", "tcp://127.0.0.1:28334"),
             ("hashblock", "tcp://127.0.0.1:28335"),
-        ])
+        ], sync_blocks=False)
```
Comment (jonatack) [inline test/functional/interface_zmq.py:520]: nit, maybe add a comment clarifying why sync_blocks must be false (the test hangs without it)

**Review by jonatack — COMMENTED** [review]
ACK 9a9453d35be5e6d24e14f75c911428a5dbbd2b30

Might be good if @instagibbs had a look.

(IDK if he prefers instagibbs or Gregory Sanders for the commit credit)

File: test/functional/interface_zmq.py (hunk 3 above)
Comment (theStack) [reply test/functional/interface_zmq.py]: Thanks, I decided for the first variant, i.e. enforcing named args for `recv_timeout` and `sync_blocks`.

File: test/functional/interface_zmq.py (hunk 4 above)
Comment (theStack) [reply test/functional/interface_zmq.py:525]: Thanks, done.

**Comment by theStack** [comment]
Force-pushed with a rebase on master and suggestions by jonatack (https://github.com/bitcoin/bitcoin/pull/21008#discussion_r566907134 and https://github.com/bitcoin/bitcoin/pull/21008#discussion_r566908151).

**Review by jonatack — COMMENTED** [review]
Light ACK ef21fb7313005a8a2d4f03fb4056f1f66c1b04f0 with the caveat that I was unable to make the test fail with valgrind both here and on master, so I can't vouch that it actually fixes the CI flakiness. The test does run ~2x faster with this.

Thanks for adding the comment. It would be good for the tests to not be order-dependent.

**Pull request merged**